### ISSUE TITLE
Personal Agent

### ISSUE THREAD

**Issue by 0x4007** [ubiquity-os/plugins-wishlist#3]
# Task

- This will be registered under `issues_comment.created`
- This will look for username tags at the beginning of any comment, and relay everything to their self hosted plugin. 
- The rest of the magic happens within their own self hosted plugin so this should be a super simple plugin to build.

### Config

The host repository name. For example:

```yml
plugins:
  - uses:
    - plugin: ubiquity-os-marketplace/ubiquity-os-agent
      with:
        target: ubiquity-os-agent
```

Then comments starting with `@0x4007` will relay the full `issues_comment.created` payload to `0x4007/ubiquity-os-agent`

# Context

This one I'm very excited about. The vision here is that we can make custom user "agents" (i.e. plugins with LLMs) that are hosted by the user's GitHub (so they can modify it) and will automate actions for the user (with their PAT to authorize as them) with the full context of a particular repository/organization.  

- We make a repository that power users are intended to fork, for example `@ubiquibot/personal-agent` -> `@pavlovcik/personal-agent`
- A repository/organization configures personal agents command to be `/@<user's name>` `/@pavlovcik` maybe something like that. This should also support arguments, for example a sentence that can be parsed by an LLM `/@pavlovcik review my pull #123`
   - This technically would allow other users to invoke other user's agents. We can easily see if the invoker is an "authorized" user by checking the event context, and hard coding authorized users (self) in the boilerplate plugin code. 
   - I wonder if it would be more useful if we just look for comments that start with a username tag instead, it might be more natural to set up automations for common questions/requests i.e. `@pavlovcik can i work on this issue?` then my agent, with my PAT, and my custom prompt saying what to do in this situation, would just automatically assign them and explain the `/start` command.
- The kernel will invoke a request (and pass all parameters) to that user's plugin/agent (hosted at `@username/personal-agent` actions)
- The user can grant access to their PAT from their agent, allowing the agent to act on behalf of the owner inheriting their permissions. 

There are some ways we can make the template code which will be forked:
1. simple starting point would be just template/boilerplate that doesn't do anything
2. code makes a call to an LLM (we could even run a small model locally on the GitHub Action runner potentially in order to make dealing with credentials/API keys more hands off, at the tradeoff of it being dumber than ChatGPT etc but decentralization/free is cool)
   1. this LLM has a big prompt in the template that explains the context (you're running in a github action runner and a user invoked you from this repo...) and its capabilities (we can provide some local functions from our SDK that it can invoke to perform specific tasks by using an authenticated octokit instance using the person's PAT. It also receives all of the context of the event invocation (which user called the function, what repository and organization is it coming in from? possibly even scraping all the linked issues and pull requests for more context)
      2. If we can reliably get the LLM to write working code with Octokit (or just raw CURLs with the PAT) then we can have a context aware and english language input to any function a user can perform on GitHub (limited to the PAT permissions) which is quite interesting. 
      3. The user can "fine-tune" their LLM by adding extra details and preferences to their prompt in their forked code. I imagine that I would continue to add new sections as I see repetitive questions/queries.

---

Assuming that the org config enables support for personal agents, technically we can extend personal agent capabilities beyond GitHub. Generic telegram example: `@pavlovcik send me the credentials on Telegram @username` with the right code in my personal agent, the GitHub Action can send information to their Telegram. All invoked from the GitHub Action runner!

This could make plugin development a lot more exciting and rapid. If the team all works on their own agents, and tests them in production, we could extract useful bits from eachothers' and release "official" plugins which may normally have slower r&d cycles. 

In the further future, our kernel can support webhooks coming in from other services (like Telegram) and invoke user agents which can be a very powerful architecture for platform composability. For example, a bot call (can be "inline" in a dm to someone as well) that will pass along the conversation context to our kernel, then to a user's personal agent (github action) back to kernel and then back to Telegram

### Notes for @pavlovcik/personal-agent

- I want to make use of the XP system (as an admin) to soft incentivize/disincentivize behaviors. 
   - Prompt follow ups: there are situations where I tag team members for input and they take days to reply. I think if they take longer than 24 hours to reply, I would want to dock XP, and include an automated follow up (perhaps even on Telegram dm!) High performing team members generally reply promptly. XP can be used as a heartbeat for how actively engaged the contributor is, and how well they are performing, which is important for performance evaluations regarding base pay. 
   - On the other end of the spectrum: unnecessary tags[^1^]. If I make it clear to team members that I am around to help but more for emergencies, I would appreciate not being pinged on things unless its essential. Would be interesting to make a personal agent that will automatically reply (like an away message) explaining this, while also scrubbing out the tag from their message. Assuming it is during my awake/working hours, I would still receive a push notification on my device from the original tag. 

### Comment rewrites:

From my phone sometimes writing comments can be arduous with the custom vocabulary we use and the autocorrect. A simple agent that will save me from a lot of frustration is to edit my comments posted, and correct any typos when I post from my phone. 

### Review and follow up:

Sometimes a pull request will be 99% of the way there. It will be something like "just make sure CI passes" or "fix merge conflicts" 

Ideally the personal agent should monitor pulls that I approved and are still opened. If I said something like this, and if those conditions are met, it should merge the pull. 

- Example: https://github.com/ubiquity-os/plugin-template/pull/23#issuecomment-2391416311

[^1^]: Although it is not clear to me how we can capture the event from this. I suppose I would need to manually add in the org/repo config for `issue_comment.created`.

###### Similar [^01^]

[^01^]: [2025 Plugins Wishlist](https://www.github.com/ubiquity-os/plugins-wishlist/issues/52) 84%
[^02^]: [New Task Or Edit Pull Arbiter](https://www.github.com/ubiquity-os/plugins-wishlist/issues/53) 82%

**Comment by 0x4007** [comment]
@gentlementlegen how would this look from a plugin config standpoint? Perhaps we make a special adapter? For example, we attach an `issues_comment.created` plugin that will route the request to one's "personal agent"?

**Comment by gentlementlegen** [comment]
If you do like other plugins and hook it to the `issue_comment.created` you will already get the full payload for that event, how would this differ?

**Comment by 0x4007** [comment]
I take that as a yes. Then we can make the only config option to be the host repository name. For example:

```yml
   - with:
      - name: ubiquity-os-agent
```

**Comment by gentlementlegen** [comment]
As long as you have the bot installed in your organization, you can hook any plugin from your organization you want to see running so
```
plugins:
  - uses: 0x4007/ubiquity-os-agent
```
would run your agent.

**Comment by 0x4007** [comment]
> This issue seems to be similar to the following issue(s):
> 
> * [2025 Plugins Wishlist](https://github.com/ubiquity-os/plugins-wishlist/issues/52) Similarity: 82.52
> * [Collaborator Gating Based On Label](https://github.com/ubiquity-os-marketplace/command-start-stop/issues/62) Similarity: 78.27
> * [Near Instant GitHub Actions Cold Boot Times](https://github.com/ubiquity-os/plugin-template/issues/24) Similarity: 76.98
> * [Do not break the loop when a plugin fails to run](https://github.com/ubiquity-os/ubiquity-os-kernel/issues/120) Similarity: 75.84

@sshivaditya2019 this is becoming recursive so I don't think its working correctly. It just keeps stacking more every time I create a new issue.

**Comment by EresDev** [comment]
> We make a repository that power users are intended to fork, for example @ubiquibot/personal-agent -> @pavlovcik/personal-agent

Who is a power user here? Are we restricting it to specific contributors or parters? It looks like anyone can host the personal-agent plugin. 


We can't have this in the configs. 
```
plugins:
  - uses: 0x4007/ubiquity-os-agent
```

The reason is if we are all going to use it, are we going to edit configs for all contributors and contributors joining us in future. 

```
plugins:
  - uses: 0x4007/ubiquity-os-agent
  - uses: EresDev/ubiquity-os-agent
  - uses: gentlementlegen/ubiquity-os-agent
  and so on....
```

So, if I am right here. We have two solutions. 

1. Let the kernel call to this personal agent plugin dynamically by looking at the comment. However, it appears to me that it adds to the responsibility of kernel, and we should not do that.
2. A second plugin, personal-agent-bridge that accepts event from kernel and calls the specific personal-agent plugin. 

So, right now, solution 2 seems right to me. Let me know your thoughts. 


As I understand, the deliverables of this issue are one or two plugins (depending on which solution we choose), and provide and show successful commication between them on correct preconditions (the `/@EresDev` mention).

**Comment by 0x4007** [comment]
2. The bridge makes sense.

**Comment by EresDev** [comment]
> ```diff
> ! EresDev you were previously unassigned from this task. You cannot be reassigned.
> ```

What to do about this now? 

I am working on it in these repos. 
https://github.com/EresDevOrg/personal-agent-bridge
https://github.com/EresDevOrg/personal-agent

**Comment by 0x4007** [comment]
Normally you're supposed to open a pull on your own repository and link it here so that we can keep track of progress. You should do that now to maintain your assignment.

**Comment by gentlementlegen** [comment]
I think it didn't pick up the pull-request because you do not target this repository / organization.

**Comment by 0x4007** [comment]
Actually it only works if you target the default branch, `development`

**Comment by EresDev** [comment]
I am looking into a way to authorize the communication between the bridge and the personal agent. The problems is, you can't run workflow on personal agent from the bridge. The reason is, bridge is going to be hosted by our organization, but the personal agent is going to be hosted by the user's account. You can't run the workflow of personal agent plugin fork from the bridge with some kind of authorization/authentication.

One way that works smoothly is, UbiquityOS is also installed by user on the personal agent repository fork.  But do we have some other concerns here? Like UbiquityOS responding to other commands in their personal agent fork? 

Other way I have in mind is to ask personal agent fork user to store their encrypted PAT in some config file in the repository that the bridge can fetch, decrypt, and use. Bridge needs this only to run the personal agents workflow, and this approach looks like an overkill to me. 

Do you have any other suggestions? I would appreciate it.

**Comment by 0x4007** [comment]
> I am looking into a way to authorize the communication between the bridge and the personal agent. The problems is, you can't run workflow on personal agent from the bridge. The reason is, bridge is going to be hosted by our organization, but the personal agent is going to be hosted by the user's account. You can't run the workflow of personal agent plugin fork from the bridge with some kind of authorization/authentication.
> 
> 
> 
> One way that works smoothly is, UbiquityOS is also installed by user on the personal agent repository fork.  

This is perfectly acceptable. 

> But do we have some other concerns here? Like UbiquityOS responding to other commands in their personal agent fork? 

It must be configured to do so with the yml file so this shouldn't happen unintentionally 
 
> 
> 
> Other way I have in mind is to ask personal agent fork user to store their encrypted PAT in some config file in the repository that the bridge can fetch, decrypt, and use. 

We have our built in encryption with the X25519 key so we can do the same if necessary. 

> Bridge needs this only to run the personal agents workflow, and this approach looks like an overkill to me. 
> 
> 
> 
> Do you have any other suggestions? I would appreciate it.

**Comment by EresDev** [comment]
> > I am looking into a way to authorize the communication between the bridge and the personal agent. The problems is, you can't run workflow on personal agent from the bridge. The reason is, bridge is going to be hosted by our organization, but the personal agent is going to be hosted by the user's account. You can't run the workflow of personal agent plugin fork from the bridge with some kind of authorization/authentication.
> > One way that works smoothly is, UbiquityOS is also installed by user on the personal agent repository fork.
> 
> This is perfectly acceptable.

I wanted to do it like this. It would have made the usage very simple, fork the personal agent and it starts working. But I have sen that the kind of auth token kernel provides doesn't trigger workflows of personal agent forks on other github accounts. Because of this, I am going with Github PAT with X25519 encryption mechanism.

**Comment by EresDev** [comment]
> @EresDev, this task has been idle for a while. Please provide an update.

Hoping to be back to this soon.

**Comment by EresDev** [comment]
> @EresDev, this task has been idle for a while. Please provide an update.

Working on it.

**Comment by Keyrxng** [comment]
> One way that works smoothly is, UbiquityOS is also installed by user on the personal agent repository fork.

I thought the official UbiquityOS app is installed via GH marketplace and is a single subscription per org, no? Maybe I've been mistaken in thinking that?

**Comment by EresDev** [comment]
> > One way that works smoothly is, UbiquityOS is also installed by user on the personal agent repository fork.
> 
> I thought the official UbiquityOS app is installed via GH marketplace and is a single subscription per org, no? Maybe I've been mistaken in thinking that?

The way I mentioned above eventually didn't work for me. It worked only in same organization, so I went with github PAT.

**Comment by EresDev** [comment]
> [@EresDev](https://github.com/EresDev), this task has been idle for a while. Please provide an update.

PR is in the process of review.

**Comment by EresDev** [comment]
> [@EresDev](https://github.com/EresDev), this task has been idle for a while. Please provide an update.

Going to work on reviews tomorrow.

**Comment by 0x4007** [comment]
@gentlementlegen disqualifier seems to repeat itself when there are multiple assignees.

**Comment by 0x4007** [comment]
> Caution
> 
> All linked pull requests must be closed to generate rewards.

Somebody should go in and change all the Cautions to Warnings unless they are real errors bubbled up (like real thrown errors with internal problems)

**Comment by whilefoo** [comment]
@EresDev can you close https://github.com/EresDevOrg/personal-agent-bridge/pull/3

**Comment by 0x4007** [comment]
> @EresDev can you close https://github.com/EresDevOrg/personal-agent-bridge/pull/3

I feel like we need a slash command to handle this edge case. I don't like the idea of even an admin (me) not being able to push this through.

**Comment by gentlementlegen** [comment]
I think it's weird that there is no way to manually unlink pull-requests from GitHub's UI, this means you can pollute any repository by linking random pull-requests. I did not find anything to counter this, so yes let's have some command.

**Comment by 0x4007** [comment]
You can also "link back" or reference other issues to pollute their timelines as well.

**Comment by EresDev** [comment]
> [@EresDev](https://github.com/EresDev) can you close [EresDevOrg/personal-agent-bridge#3](https://github.com/EresDevOrg/personal-agent-bridge/pull/3)

@whilefoo  Done.

**Comment by 0x4007** [comment]
Hmm something tells me this task time was under estimated. 

We should have a protocol for this that is trust minimized. 

The labels are used as training data for the fine tuned model that will be automatically setting labels for us.

**Comment by whilefoo** [comment]
> I think it's weird that there is no way to manually unlink pull-requests from GitHub's UI, this means you can pollute any repository by linking random pull-requests. I did not find anything to counter this, so yes let's have some command.

We can modify the plugin to ignore PRs that originate from other organizations, because we can't close them (we can only close PRs from ubiquity orgs). 

The only problem is how to handle cross-org, for example the issue is in `plugins-wishlist` which is in `ubiquity-os` but the PR will be in `ubiquity-os-marketplace` and similar scenarios, so you'd need to specify which orgs are allowed in the config.
```yml
allowedPrOrganizations:
  - ubiquity-os
  - ubiquity-os-marketplace
```

> Hmm something tells me this task time was under estimated.
> 
> We should have a protocol for this that is trust minimized.
> 
> The labels are used as training data for the fine tuned model that will be automatically setting labels for us.

We could have a command that the assignees would trigger after the issue is closed and tell how long they spent on the task so we'd have data for training the model, but it relies on trust (maybe only for trusted contributors?)

**Comment by 0x4007** [comment]
> We could have a command that the assignees would trigger after the issue is closed and tell how long they spent on the task so we'd have data for training the model, but it relies on trust (maybe only for trusted contributors?)

There isn't an incentive to lie. It's an interesting idea for collaborators only!

**Comment by whilefoo** [comment]
> Instead of having to specify organizations in the list, maybe we could just check if the bot has access during the run?

But if the bot has access, it doesn't mean we have access to close it, unless you mean to ignore PRs that bot doesn't have access to and make a command to close PRs

**Comment by 0x4007** [comment]
> > Instead of having to specify organizations in the list, maybe we could just check if the bot has access during the run?
> 
> But if the bot has access, it doesn't mean we have access to close it, 

It can because we already see this happen with disqualifier. 

> unless you mean to ignore PRs that bot doesn't have access to and make a command to close PRs

I think this makes sense, yes. 

As a side note I expect that most pulls the bot might not have access to, like envision a new contributor. It's unlikely they installed the bot. 

But it would look the best if we see the bot close everything out before distributing rewards.

**Comment by whilefoo** [comment]
> As a side note I expect that most pulls the bot might not have access to, like envision a new contributor. It's unlikely they installed the bot.

If the PR targets ubiquity's repository then the PR is owned by that repository so the bot will have access. The only time the bot doesn't have access is when they open a PR targeting their own fork like https://github.com/EresDevOrg/personal-agent-bridge/pull/3 but in normal use that shouldn't happen

**Comment by 0x4007** [comment]
I think if their fork is targeting our repository we don't own the pull and can't admin it.

**Comment by 0x4007** [comment]
> > Hmm something tells me this task time was under estimated.
> > 
> > We should have a protocol for this that is trust minimized.
> > 
> > The labels are used as training data for the fine tuned model that will be automatically setting labels for us.
> 
> We could have a command that the assignees would trigger after the issue is closed and tell how long they spent on the task so we'd have data for training the model, but it relies on trust (maybe only for trusted contributors?)
> 
> 

For now you can post here how long this should have taken for reference.

**Comment by whilefoo** [comment]
> I think if their fork is targeting our repository we don't own the pull and can't admin it.

What do you mean? Our flow is we fork the repo, commit changes in a new branch and make a PR against the original repo - the pull will be owned by original repo

**Comment by 0x4007** [comment]
/ask what do you think is the correct answer?

**Comment by 0x4007** [comment]
We definitely should increase focus on the conversation and far less on repository details @shiv810

**Comment by 0x4007** [comment]
/ask do you agree or disagree with my comment "I think if their fork is targeting our repository we don't own the pull and can't admin it." in the context of this discussion?

**Comment by 0x4007** [comment]
@whilefoo could you help me make a personal agent that will rewrite and correct typos from mobile? I can possibly drop in a signature when writing from mobile like a line break at the end (the keyboard on iOS always covers the bottom lines of my draft)

**Comment by whilefoo** [comment]
/ask if someone forks our repository, commits changes in a new branch and makes a pull request with base branch being our repository and head branch being the fork, is the pull request considered a part of our repository or a part of the fork. In other words do we have permission to close the PR?

**Comment by 0x4007** [comment]
It only rewrites when I double line break at the end
